Index: exams/answers.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exams/answers.json b/exams/answers.json
new file mode 100644
--- /dev/null	(date 1728116733945)
+++ b/exams/answers.json	(date 1728116733945)
@@ -0,0 +1,9 @@
+{
+  "EGE": {
+    "1": [34, 35, 37, 34, 32, 53, 45],
+    "2": ["xwzy", "zywx", "xyzw", "yzxw", "xzyw" , "xzyw", "ywzx"],
+    "3": [1, 2, 3, 4, 5 , 6, 7],
+    "4": [8, 9, 12, 13, 18, 14, 18],
+    "5": [61, 89, 66, 68, 9, 25, 12]
+  }
+}
Index: MediaGroupHandler.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/MediaGroupHandler.py b/MediaGroupHandler.py
new file mode 100644
--- /dev/null	(date 1727965864400)
+++ b/MediaGroupHandler.py	(date 1727965864400)
@@ -0,0 +1,59 @@
+import asyncio
+from typing import Any, Dict, Union
+
+from aiogram import BaseMiddleware
+from aiogram.types import Message
+
+
+class AlbumMiddleware(BaseMiddleware):
+    def __init__(self, latency: Union[int, float] = 0.1):
+        # Initialize latency and album_data dictionary
+        self.latency = latency
+        self.album_data = {}
+#
+    def collect_album_messages(self, event: Message):
+        """
+        Collect messages of the same media group.
+        """
+#         # Check if media_group_id exists in album_data
+        if event.media_group_id not in self.album_data:
+#             # Create a new entry for the media group
+            self.album_data[event.media_group_id] = {"messages": []}
+#
+#         # Append the new message to the media group
+        self.album_data[event.media_group_id]["messages"].append(event)
+#
+#         # Return the total number of messages in the current media group
+        return len(self.album_data[event.media_group_id]["messages"])
+#
+    async def __call__(self, handler, event: Message, data: Dict[str, Any]) -> Any:
+        """
+        Main middleware logic.
+        """
+#         # If the event has no media_group_id, pass it to the handler immediately
+        if not event.media_group_id:
+            return await handler(event, data)
+#
+#         # Collect messages of the same media group
+        total_before = self.collect_album_messages(event)
+#
+#         # Wait for a specified latency period
+        await asyncio.sleep(self.latency)
+#
+#         # Check the total number of messages after the latency
+        total_after = len(self.album_data[event.media_group_id]["messages"])
+#
+#         # If new messages were added during the latency, exit
+        if total_before != total_after:
+            return
+#
+#         # Sort the album messages by message_id and add to data
+        album_messages = self.album_data[event.media_group_id]["messages"]
+        album_messages.sort(key=lambda x: x.message_id)
+        data["album"] = album_messages
+#
+#         # Remove the media group from tracking to free up memory
+        del self.album_data[event.media_group_id]
+#         # Call the original event handler
+        return await handler(event, data)
+#
\ No newline at end of file
Index: functions.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/functions.py b/functions.py
new file mode 100644
--- /dev/null	(date 1728034029919)
+++ b/functions.py	(date 1728034029919)
@@ -0,0 +1,23 @@
+from aiogram import Bot, Dispatcher, types, F
+from aiogram.utils.keyboard import InlineKeyboardBuilder
+
+class Functions:
+    @staticmethod
+    async def show_students(USER_DICT: dict ,callback_text: str, message_text: str, callback: types.CallbackQuery):
+        builder = InlineKeyboardBuilder()
+        for user_id in USER_DICT.keys():
+            builder.add(types.InlineKeyboardButton(text=f'{USER_DICT[user_id]['name']}',
+                                                   callback_data=f'{callback_text}{user_id}'))
+        kb = builder.adjust(3).as_markup(resize_keyboard=True)
+        await callback.message.answer(f"{message_text}", reply_markup=kb)
+
+    @staticmethod
+    def translate_num_to_rus(number: int):
+        if 5 <= number % 100 <= 20:
+            return 'баллов'
+        elif number % 10 == 1:
+            return 'балл'
+        elif number % 10 in [2, 3, 4]:
+            return 'балла'
+        else:
+            return 'баллов'
\ No newline at end of file
Index: server.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/server.py b/server.py
new file mode 100644
--- /dev/null	(date 1728117037700)
+++ b/server.py	(date 1728117037700)
@@ -0,0 +1,499 @@
+import asyncio
+import os
+from datetime import date
+import json
+from aiogram.enums.parse_mode import ParseMode
+
+from aiogram import Bot, Dispatcher, types, F
+from aiogram.filters.command import Command
+from aiogram.fsm.context import FSMContext
+from aiogram.fsm.state import State, StatesGroup
+from aiogram.types import CallbackQuery, InputFile, FSInputFile, InputMediaPhoto, InlineKeyboardMarkup, \
+    InlineKeyboardButton
+from aiogram.utils.keyboard import InlineKeyboardBuilder
+import random
+from functions import Functions
+from MediaGroupHandler import AlbumMiddleware
+from functions import Functions
+
+# Объект бота
+bot = Bot(token="7615724794:AAHst5y-K5F5VGLAyjxwLIhih7pCvpJakaU")
+# Диспетчер
+dp = Dispatcher()
+ANSWER = 0
+ADMIN_ID = 355721034
+username = ""
+FILE_NAME = ''
+USER_ID_DOWNLOAD = 0
+USER_ID_POINTS = 0
+MEDIA_GROUP = []
+dp.message.middleware(AlbumMiddleware())
+PATH = ''
+MSG = None
+USER_ID = 0
+SIZES = ()
+DATA = ''
+class States(StatesGroup):
+    STATE_register = State()
+    STATE_file = State()
+    STATE_points = State()
+    STATE_homework = State()
+    STATE_get_media_group = State()
+    STATE_send_exam = State()
+    STATE_add_date = State()
+    STATE_edit_date = State()
+
+if os.path.exists('data/user_data.json'):
+    with open('data/user_data.json') as f:
+        USER_DICT = json.load(f)
+    print(USER_DICT)
+else:
+    USER_DICT = {}
+WAYS = ['ЕГЭ - Информатика', 'ОГЭ - Информатика', 'ОГЭ - Математика', 'Информатика', 'Математика']
+def save_to_file():
+    js = json.dumps(USER_DICT)
+    with open('data/user_data.json', 'w') as f:
+        f.write(js)
+    print(USER_DICT)
+@dp.message(F.text == "Регистрация")
+async def cmd_register(message: types.Message, state: FSMContext):
+    await message.answer(f"Введи своё имя")
+    await state.set_state(States.STATE_register)
+
+
+@dp.message(States.STATE_register)
+async def get_data(message: types.Message, state: FSMContext):
+    user_id = str(message.from_user.id)
+    if user_id not in USER_DICT:
+        USER_DICT[user_id] = {}
+
+    if "name" not in USER_DICT[user_id]:
+        USER_DICT[user_id]["name"] = message.text
+        await message.answer(f'Привет, {message.text}. Как твоя фамилия?')
+    elif "last_name" not in USER_DICT[user_id]:
+        USER_DICT[user_id]["last_name"] = message.text
+        b1 = types.KeyboardButton(text='ЕГЭ - Информатика')
+        b2 = types.KeyboardButton(text='ОГЭ - Информатика')
+        b3 = types.KeyboardButton(text='ОГЭ - Математика')
+        b4 = types.KeyboardButton(text='Информатика')
+        b5 = types.KeyboardButton(text='Математика')
+        kb = types.ReplyKeyboardMarkup(keyboard=[[b1], [b2], [b3], [b4, b5]], resize_keyboard=True, one_time_keyboard=True)
+        await message.answer(f"Направление подготовки?", reply_markup=kb)
+    elif 'way' not in USER_DICT[user_id]:
+        if message.text not in WAYS:
+            b1 = types.KeyboardButton(text='ЕГЭ - Информатика')
+            b2 = types.KeyboardButton(text='ОГЭ - Информатика')
+            b3 = types.KeyboardButton(text='ОГЭ - Математика')
+            b4 = types.KeyboardButton(text='Информатика')
+            b5 = types.KeyboardButton(text='Математика')
+            kb = types.ReplyKeyboardMarkup(keyboard=[[b1], [b2], [b3], [b4, b5]], resize_keyboard=True,
+                                           one_time_keyboard=True)
+            await message.answer(f"Направление подготовки?", reply_markup=kb)
+        else:
+            USER_DICT[user_id]['way'] = message.text
+            path = ''
+            if message.text == 'ЕГЭ - Информатика':
+               path = f'Homework/Ege-IT/{USER_DICT[user_id]['name']}-{USER_DICT[user_id]['last_name']}'
+            elif message.text == 'ОГЭ - Информатика':
+                path = f'Homework/Oge-IT/{USER_DICT[user_id]['name']}-{USER_DICT[user_id]['last_name']}'
+            elif message.text == 'ОГЭ - Математика':
+                path = f'Homework/Oge-Math/{USER_DICT[user_id]['name']}-{USER_DICT[user_id]['last_name']}'
+            elif message.text == 'Информатика':
+                path = f'Homework/IT/{USER_DICT[user_id]['name']}-{USER_DICT[user_id]['last_name']}'
+            elif message.text == 'Математика':
+                path = f'Homework/Math/{USER_DICT[user_id]['name']}-{USER_DICT[user_id]['last_name']}'
+            if not os.path.exists(path):
+                os.mkdir(path)
+            save_to_file()
+            await message.answer(f"Спасибо за регистрацию!")
+            await state.clear()
+            await cmd_menu(message)
+
+
+def return_folder_name(way: str):
+    if way == 'ЕГЭ - Информатика':
+        return 'Ege-IT'
+    elif way == 'ОГЭ - Информатика':
+        return 'Oge-IT'
+    elif way == 'ОГЭ - Математика':
+        return 'Oge-Math'
+    elif way == 'Информатика':
+        return 'IT'
+    elif way == 'Математика':
+        return 'Math'
+# Хэндлер на команду /start
+@dp.message(Command("start"))
+async def cmd_start(message: types.Message):
+    await message.answer(f"Привет, {message.from_user.username}")
+    await cmd_menu(message)
+
+
+@dp.message(Command("menu"))
+async def cmd_menu(message: types.Message):
+    if message.from_user.id == ADMIN_ID:
+        b1 = types.InlineKeyboardButton(text="Отправить домашнее задание", callback_data='send-homework')
+        b2 = types.InlineKeyboardButton(text="Добавить задание на ОГЭ / ЕГЭ", callback_data='add-exam')
+        b3 = types.InlineKeyboardButton(text='Показать статистику', callback_data='send-statistic')
+        b4 = types.InlineKeyboardButton(text='Добавить баллы', callback_data='add_points')
+        b5 = types.InlineKeyboardButton(text='Поставить расписание', callback_data='add_dates')
+        b6 = types.InlineKeyboardButton(text='Изменить расписание', callback_data='change_dates')
+        kb = types.InlineKeyboardMarkup(inline_keyboard=[[b1], [b2], [b3, b4], [b5, b6]])
+        await message.answer(f"Что делаем?", reply_markup=kb)
+    elif str(message.from_user.id) not in USER_DICT.keys():
+        b1 = types.KeyboardButton(text="Регистрация")
+        kb = types.ReplyKeyboardMarkup(keyboard=[[b1]], resize_keyboard=True, one_time_keyboard=True)
+        await message.answer(f"Пожалуйста, зарегистрируйся.", reply_markup=kb)
+    else:
+        b1 = types.KeyboardButton(text="Домашнее задание")
+        b2 = types.KeyboardButton(text="Баллы за выполнение домашнего задания")
+        b3 = types.KeyboardButton(text='Статистика')
+        b4 = types.KeyboardButton(text='Сдать домашнее задание')
+        b5 = types.KeyboardButton(text='Моё расписание')
+        if USER_DICT[str(message.from_user.id)]['way'] in ['ЕГЭ - Информатика', 'ОГЭ - Информатика', 'ОГЭ - Математика']:
+            b6 = types.KeyboardButton(text='Решать экзаменационные задачи')
+            kb = types.ReplyKeyboardMarkup(keyboard=[[b1, b3],[b2], [b4, b5], [b6]], resize_keyboard=True, one_time_keyboard=True)
+        else:
+            kb = types.ReplyKeyboardMarkup(keyboard=[[b1, b3], [b2], [b4, b5]], resize_keyboard=True,
+                                           one_time_keyboard=True)
+        await message.answer(f"Чем бы ты хотел заняться?", reply_markup=kb)
+
+@dp.message(F.text == "Моё расписание")
+async def return_schedule(message: types.Message):
+    if 'dates' in USER_DICT[str(message.from_user.id)].keys():
+        answer = (f'<b>{USER_DICT[str(message.from_user.id)]['name']} {USER_DICT[str(message.from_user.id)]['last_name']}</b>\n'
+                  f'Твоё расписание (время указано по Москве):\n\n')
+        for data in USER_DICT[str(message.from_user.id)]['dates']:
+            answer += f'{data}\n'
+    else:
+        answer = 'Тебе пока не поставили расписание'
+    await message.answer(answer, parse_mode=ParseMode.HTML)
+
+@dp.callback_query(F.data == "send-homework")
+async def send_homework(callback: types.CallbackQuery):
+    await Functions.show_students(USER_DICT, 'student_homework_', 'Выбери студента', callback)
+    await callback.answer()
+
+@dp.callback_query(F.data == "add_points")
+async def add_points(callback: types.CallbackQuery):
+    await Functions.show_students(USER_DICT, 'student_points_', 'Выбери студента', callback)
+
+@dp.callback_query(F.data.startswith('student_points_'))
+async def cmd_selected_student_points(call: CallbackQuery, state: FSMContext):
+    global USER_ID_POINTS
+    USER_ID_POINTS = str(call.data).replace('student_points_', '')
+    await call.message.edit_text(f'Сколько баллов добавить? Если нужно убавить, отправьте знак "-" перед числом')
+    await state.set_state(States.STATE_points)
+    await call.answer()
+
+@dp.callback_query(F.data.startswith('student_homework_'))
+async def cmd_selected_student_homework(call: CallbackQuery, state: FSMContext):
+    global USER_ID_DOWNLOAD
+    USER_ID_DOWNLOAD = str(call.data).replace('student_homework_', '')
+    await call.message.edit_text(f'Отправьте файл домашнего задания')
+    await state.set_state(States.STATE_file)
+    await call.answer()
+
+@dp.message(States.STATE_points)
+async def set_points(message: types.Message, state: FSMContext):
+    if 'points' not in USER_DICT[USER_ID_POINTS].keys():
+        USER_DICT[USER_ID_POINTS]['points'] = 0
+    try:
+        USER_DICT[USER_ID_POINTS]['points'] += int(message.text)
+        await message.answer(f'Студенту {USER_DICT[USER_ID_POINTS]['name']} {USER_DICT[USER_ID_POINTS]['last_name']} добавлено {message.text} баллов.\n'
+                            f'Количество баллов: {USER_DICT[USER_ID_POINTS]['points']}')
+        if int(message.text) > 0:
+            await bot.send_message(int(USER_ID_POINTS), f'Тебе добавили {message.text} {Functions.translate_num_to_rus(abs(int(message.text)))}\n'
+                                                        f'Твой баланс: {USER_DICT[USER_ID_POINTS]['points']}')
+        else:
+            await bot.send_message(int(USER_ID_POINTS), f'У тебя забрали {abs(int(message.text))} {Functions.translate_num_to_rus(abs(int(message.text)))}\n'
+                                                        f'Твой баланс: {USER_DICT[USER_ID_POINTS]['points']}')
+        save_to_file()
+        await state.clear()
+    except:
+        await message.answer(text='Вы ввели некорректное число. Попробуйте ещё раз')
+
+
+@dp.message(States.STATE_file)
+async def get_file(message: types.Message, state: FSMContext):
+    if 'count_homeworks' not in USER_DICT[USER_ID_DOWNLOAD].keys():
+        USER_DICT[USER_ID_DOWNLOAD]['count_homeworks'] = 1
+    else:
+        USER_DICT[USER_ID_DOWNLOAD]['count_homeworks'] += 1
+
+    if 'total_homeworks' not in USER_DICT[USER_ID_DOWNLOAD].keys():
+        USER_DICT[USER_ID_DOWNLOAD]['total_homeworks'] = 0
+
+    path = f'Homework/{return_folder_name(USER_DICT[USER_ID_DOWNLOAD]['way'])}/{USER_DICT[USER_ID_DOWNLOAD]['name']}-{USER_DICT[USER_ID_DOWNLOAD]['last_name']}/{date.today()}_{USER_DICT[USER_ID_DOWNLOAD]['count_homeworks']}.pdf'
+    await message.bot.download(file=message.document.file_id, destination=path)
+    await message.answer(text=f'Домашнее задание отправленно!')
+    file = FSInputFile(path=path)
+    await bot.send_document(int(USER_ID_DOWNLOAD), file, caption='Тебе пришло новое домашнее задание!')
+    save_to_file()
+    await state.clear()
+
+@dp.callback_query(F.data == "start")
+async def cb_menu(callback: types.CallbackQuery):
+    await cmd_menu(callback.message)
+    await callback.answer()
+
+@dp.message(F.text == "Домашнее задание")
+async def send_homework_for_user(message: types.Message):
+    path = f'Homework/{return_folder_name(USER_DICT[str(message.from_user.id)]['way'])}/{USER_DICT[str(message.from_user.id)]['name']}-{USER_DICT[str(message.from_user.id)]['last_name']}'
+    files = os.listdir(path)
+    if len(files) == 0:
+        await message.answer('У тебя нет домашнего задания')
+    else:
+        for file in files:
+            file_to_send = FSInputFile(path=path + '/' + file)
+            await bot.send_document(int(str(message.from_user.id)), file_to_send)
+
+@dp.message(F.text == "Сдать домашнее задание")
+async def send_homework_for_user(message: types.Message):
+    path = f'Homework/{return_folder_name(USER_DICT[str(message.from_user.id)]['way'])}/{USER_DICT[str(message.from_user.id)]['name']}-{USER_DICT[str(message.from_user.id)]['last_name']}'
+    files = os.listdir(path)
+    if len(files) == 0:
+        await message.answer('У тебя нет домашнего задания')
+    else:
+        builder = InlineKeyboardBuilder()
+        for file in files:
+            builder.add(types.InlineKeyboardButton(text=f'{file}',
+                                                   callback_data=f'send_to_admin_{file}'))
+        kb = builder.adjust(1).as_markup(resize_keyboard=True)
+        await message.answer('Выбери домашнее задание, которое хочешь сдать:', reply_markup=kb)
+
+@dp.callback_query(F.data.startswith('send_to_admin_'))
+async def cmd_selected_student_points(call: CallbackQuery, state: FSMContext):
+    global FILE_NAME
+    FILE_NAME = str(call.data).replace('send_to_admin_', '')
+    await call.message.edit_text(f'Отправь, пожалуйста, домашнее задание в виде фотографий')
+    await state.set_state(States.STATE_get_media_group)
+    await call.answer()
+
+@dp.message(States.STATE_get_media_group)
+async def get_media(message: types.Message, state: FSMContext, album: list = None):
+    global ADMIN_ID, FILE_NAME
+    counter = 0
+    caption = (
+        f'Студент {USER_DICT[str(message.from_user.id)]['name']} {USER_DICT[str(message.from_user.id)]['last_name']} прислал тебе решение домашнего задания\n'
+        f'{FILE_NAME}')
+    if album is not None:
+        for photo in album:
+            await photo.bot.download(file=photo.photo[-1].file_id, destination=f'Answer/p{counter}.jpg')
+            counter += 1
+        images = []
+        images_names = os.listdir('Answer')
+        i = 0
+        for photo in images_names:
+            if i == 0:
+                images.append(InputMediaPhoto(type='photo', media=FSInputFile(path=f'Answer/{photo}'), caption=caption))
+            else:
+                images.append(InputMediaPhoto(type='photo', media=FSInputFile(path=f'Answer/{photo}')))
+            i += 1
+        await bot.send_media_group(ADMIN_ID, media=images)
+        for photo in images_names:
+            if os.path.exists(f'Answer/{photo}'):
+                os.remove(f'Answer/{photo}')
+    else:
+        await message.bot.download(file=message.photo[-1].file_id, destination=f'Answer/p1.jpg')
+        await bot.send_photo(chat_id=ADMIN_ID, photo=FSInputFile(f'Answer/p{1}.jpg', 'hw'), caption=caption)
+        if os.path.exists(f'Answer/p1.jpg'):
+            os.remove(f'Answer/p1.jpg')
+
+    await message.answer('Домашнее задание отправлено!')
+    path = f'Homework/{return_folder_name(USER_DICT[str(message.from_user.id)]['way'])}/{USER_DICT[str(message.from_user.id)]['name']}-{USER_DICT[str(message.from_user.id)]['last_name']}/{FILE_NAME}'
+    if os.path.exists(path):
+        os.remove(path)
+    USER_DICT[str(message.from_user.id)]['count_homeworks'] -= 1
+    USER_DICT[str(message.from_user.id)]['total_homeworks'] += 1
+    await state.clear()
+
+@dp.message(F.text == "Баллы за выполнение домашнего задания")
+async def send_points(message: types.Message):
+    if 'points' not in USER_DICT[str(message.from_user.id)].keys():
+        USER_DICT[str(message.from_user.id)]['points'] = 0
+    await message.answer(f'Твои баллы: {USER_DICT[str(message.from_user.id)]['points']}')
+    save_to_file()
+
+@dp.message(F.text == "Статистика")
+async def statics(message: types.Message):
+    if 'points' not in USER_DICT[str(message.from_user.id)].keys():
+        USER_DICT[str(message.from_user.id)]['points'] = 0
+
+    if 'count_homeworks' not in USER_DICT[str(message.from_user.id)].keys():
+        USER_DICT[str(message.from_user.id)]['count_homeworks'] = 0
+
+    if 'total_homeworks' not in USER_DICT[str(message.from_user.id)].keys():
+        USER_DICT[str(message.from_user.id)]['total_homeworks'] = 0
+    data = USER_DICT[str(message.from_user.id)]
+    answer = (f'<b>{data['name']} {data['last_name']}</b>\n'
+              f'Твой баланс: <b>{data['points']}</b>\n'
+              f'Количество нерешенных домашних заданий: <b>{data['count_homeworks']}</b>\n'
+              f'Количество решенных домашних заданий за всё время: <b>{data['total_homeworks']}</b>')
+    await message.answer(answer, parse_mode=ParseMode.HTML)
+    save_to_file()
+
+@dp.message(F.text == "Решать экзаменационные задачи")
+async def exams(message: types.Message):
+    global PATH, USER_ID, SIZES
+    USER_ID = message.from_user.id
+    builder = InlineKeyboardBuilder()
+    user = USER_DICT[str(message.from_user.id)]
+    if 'picked_tasks' not in USER_DICT[str(message.from_user.id)].keys():
+        USER_DICT[str(message.from_user.id)]['picked_tasks'] = []
+
+    if user['way'] == 'ЕГЭ - Информатика':
+        rng = range(1, 24)
+        PATH = 'exams/EGE'
+        SIZES = (5, 5, 5, 5, 3, 1, 1)
+    else:
+        rng = range(1, 14)
+        PATH = 'exams/OGE'
+        SIZES = (5, 5, 3, 1, 1)
+    for task in rng:
+        builder.add(types.InlineKeyboardButton(text=f'{str(task)}',
+                                               callback_data=f'task_{str(task)}'))
+
+    builder.button(text=f'Начать решать',callback_data=f'continue')
+    builder.button(text=f'Закончить', callback_data=f'end')
+    kb = builder.adjust(*SIZES).as_markup(resize_keyboard=True)
+    await message.answer(f"Твой профиль: {user['way']}.\nВыбери задания для решения", reply_markup=kb)
+
+@dp.callback_query(F.data.startswith('task_'))
+async def multichoice(call: CallbackQuery):
+    global USER_ID, SIZES
+    pick = str(call.data).replace('task_', '')
+    if int(pick) in USER_DICT[str(USER_ID)]['picked_tasks']:
+        USER_DICT[str(USER_ID)]['picked_tasks'].remove(int(pick))
+    else:
+        USER_DICT[str(USER_ID)]['picked_tasks'].append(int(pick))
+    user = USER_DICT[str(USER_ID)]
+    if user['way'] == 'ЕГЭ - Информатика':
+        rng = range(1, 24)
+    else:
+        rng = range(1, 14)
+    builder = InlineKeyboardBuilder()
+    for task in rng:
+        if task in user['picked_tasks']:
+            builder.add(types.InlineKeyboardButton(text=f'{str(task)} ✅',
+                                                callback_data=f'task_{str(task)}'))
+        else:
+            builder.add(types.InlineKeyboardButton(text=f'{str(task)}',
+                                                   callback_data=f'task_{str(task)}'))
+    builder.add(types.InlineKeyboardButton(text=f'Начать решать',
+                                           callback_data=f'continue'))
+    builder.button(text=f'Закончить', callback_data=f'end')
+    kb = builder.adjust(*SIZES).as_markup(resize_keyboard=True)
+    await call.message.edit_text(f"Твой профиль: {user['way']}.\nВыбери задания для решения", reply_markup=kb)
+    await call.answer()
+
+@dp.callback_query(F.data.startswith('continue'))
+async def tasks(call: CallbackQuery, state: FSMContext):
+    global USER_ID, PATH, USER_DICT, ANSWER, MSG
+    rnd_num_task = random.choice(USER_DICT[str(USER_ID)]['picked_tasks'])
+    rnd_task = random.choice(os.listdir(f'{PATH}/{rnd_num_task}'))
+    print(rnd_task)
+    with open(f'exams/answers.json') as f:
+        print(PATH.split('/')[1], str(rnd_num_task).split('.')[0], int(str(rnd_task).split('.')[0]) - 1)
+        ANSWER = json.load(f)[PATH.split('/')[1]][str(rnd_num_task).split('.')[0]][int(str(rnd_task).split('.')[0]) - 1]
+    b1 = InlineKeyboardButton(text='Продолжить', callback_data='continue')
+    b2 = InlineKeyboardButton(text='Закончить', callback_data='end')
+    kb = InlineKeyboardMarkup(inline_keyboard=[[b1, b2]])
+    await call.message.delete()
+    MSG = await call.message.answer_photo(photo=FSInputFile(f'{PATH}/{rnd_num_task}/{rnd_task}', filename='task'), caption='Попробуй решить и присылай решение', reply_markup=kb)
+    await state.set_state(States.STATE_send_exam)
+    await call.answer()
+
+@dp.message(States.STATE_send_exam)
+async def get_answer(message: types.Message, state: FSMContext):
+    global USER_ID, PATH, USER_DICT, ANSWER
+    b1 = InlineKeyboardButton(text='Продолжить', callback_data='continue')
+    b2 = InlineKeyboardButton(text='Закончить', callback_data='end')
+    kb = InlineKeyboardMarkup(inline_keyboard=[[b1, b2]])
+    if message.text == str(ANSWER):
+        USER_DICT[str(USER_ID)]['points'] += 1
+        await MSG.delete()
+        await message.answer(f'Ты правильно решил задачу!\nТвой баланс: {USER_DICT[str(USER_ID)]['points']}', reply_markup=kb)
+        save_to_file()
+    else:
+        await MSG.delete()
+        await message.answer(f'Ты неправильно решил задачу!', reply_markup=kb)
+    await state.clear()
+
+@dp.callback_query(F.data.startswith('end'))
+async def tasks(call: CallbackQuery):
+    global USER_ID, PATH, USER_DICT
+    if str(USER_ID) in USER_DICT.keys():
+        USER_DICT[str(USER_ID)]['picked_tasks'] = []
+    USER_ID = 0
+    PATH = ''
+    await call.message.delete()
+    save_to_file()
+    await call.answer()
+
+@dp.callback_query(F.data == "add_dates")
+async def dates(callback: types.CallbackQuery):
+    await Functions.show_students(USER_DICT, 'dates_', 'Выбери студента', callback)
+    await callback.answer()
+
+@dp.callback_query(F.data.startswith('dates_'))
+async def dates_send(call: CallbackQuery, state: FSMContext):
+    global USER_ID
+    USER_ID = str(call.data).replace('dates_', '')
+    USER_DICT[USER_ID]['dates'] = []
+    await call.message.answer(f"Введите расписание в формате <День недели>:<Ч:М (мск)>\n"
+                              f"Каждую дату вводите с новой строки")
+    await state.set_state(States.STATE_add_date)
+    await call.answer()
+
+@dp.message(States.STATE_add_date)
+async def get_dates(message: types.Message, state: FSMContext):
+    schedule = str(message.text).split('\n')
+    for data in schedule:
+        USER_DICT[USER_ID]['dates'].append(data)
+    await message.answer(f"Расписание добавлено!")
+    await state.clear()
+    save_to_file()
+
+
+@dp.callback_query(F.data == "change_dates")
+async def change_dates(callback: types.CallbackQuery):
+    await Functions.show_students(USER_DICT, 'student_change_', 'Выбери студента', callback)
+    await callback.answer()
+
+@dp.callback_query(F.data.startswith('student_change_'))
+async def change_dates_select_date(call: CallbackQuery):
+    global USER_ID
+    USER_ID = str(call.data).replace('student_change_', '')
+    builder = InlineKeyboardBuilder()
+    for date_to_change in USER_DICT[USER_ID]['dates']:
+        builder.add(types.InlineKeyboardButton(text=f'{date_to_change}', callback_data=f'change_dates_{date_to_change}'))
+    kb = builder.adjust(3).as_markup(resize_keyboard=True)
+    await call.message.answer(f"Выбери дату", reply_markup=kb)
+    await call.answer()
+
+@dp.callback_query(F.data.startswith('change_dates_'))
+async def change_dates_in_dict(call: CallbackQuery, state: FSMContext):
+    global USER_ID, DATA
+    DATA = str(call.data).replace('change_dates_', '')
+    await call.message.answer(f"Введите новую дату и время в формате: <День недели>:<Ч:М (мск)>")
+    await state.set_state(States.STATE_edit_date)
+    await call.answer()
+
+@dp.message(States.STATE_edit_date)
+async def final_edit_dates(message: types.Message, state: FSMContext):
+    schedule = str(message.text)
+    ind = USER_DICT[USER_ID]['dates'].index(DATA)
+    USER_DICT[USER_ID]['dates'][ind] = schedule
+    await message.answer(f"Расписание изменено!")
+    await state.clear()
+    await bot.send_message(chat_id=USER_ID, text=f'Ваше расписание изменено!\n'
+                                                 f'{DATA} --> {schedule}')
+    save_to_file()
+
+# Запуск процесса поллинга новых апдейтов
+async def main():
+    await dp.start_polling(bot)
+
+
+if __name__ == "__main__":
+    asyncio.run(main())
Index: .idea/inspectionProfiles/Project_Default.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/inspectionProfiles/Project_Default.xml b/.idea/inspectionProfiles/Project_Default.xml
new file mode 100644
--- /dev/null	(date 1727950372920)
+++ b/.idea/inspectionProfiles/Project_Default.xml	(date 1727950372920)
@@ -0,0 +1,19 @@
+<component name="InspectionProjectProfileManager">
+  <profile version="1.0">
+    <option name="myName" value="Project Default" />
+    <inspection_tool class="PyPep8NamingInspection" enabled="true" level="WEAK WARNING" enabled_by_default="true">
+      <option name="ignoredErrors">
+        <list>
+          <option value="N806" />
+        </list>
+      </option>
+    </inspection_tool>
+    <inspection_tool class="PyUnresolvedReferencesInspection" enabled="true" level="WARNING" enabled_by_default="true">
+      <option name="ignoredIdentifiers">
+        <list>
+          <option value="_collections_abc.dict_keys.*" />
+        </list>
+      </option>
+    </inspection_tool>
+  </profile>
+</component>
\ No newline at end of file
Index: .idea/.gitignore
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/.gitignore b/.idea/.gitignore
new file mode 100644
--- /dev/null	(date 1727950373220)
+++ b/.idea/.gitignore	(date 1727950373220)
@@ -0,0 +1,3 @@
+# Default ignored files
+/shelf/
+/workspace.xml
Index: .idea/Homework Sender.iml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/Homework Sender.iml b/.idea/Homework Sender.iml
new file mode 100644
--- /dev/null	(date 1727950373230)
+++ b/.idea/Homework Sender.iml	(date 1727950373230)
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module type="PYTHON_MODULE" version="4">
+  <component name="NewModuleRootManager">
+    <content url="file://$MODULE_DIR$">
+      <excludeFolder url="file://$MODULE_DIR$/.venv" />
+    </content>
+    <orderEntry type="inheritedJdk" />
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+</module>
\ No newline at end of file
Index: data/user_data.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/data/user_data.json b/data/user_data.json
new file mode 100644
--- /dev/null	(date 1728117078610)
+++ b/data/user_data.json	(date 1728117078610)
@@ -0,0 +1,1 @@
+{"6104237517": {"name": "\u041a\u0438\u0440\u0438\u043b\u043b", "last_name": "\u0427\u0435\u043b\u044b\u0448\u0435\u0432", "way": "\u0415\u0413\u042d - \u0418\u043d\u0444\u043e\u0440\u043c\u0430\u0442\u0438\u043a\u0430", "points": 0, "count_homeworks": 0, "total_homeworks": 0, "picked_tasks": []}}
\ No newline at end of file
Index: .idea/modules.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/modules.xml b/.idea/modules.xml
new file mode 100644
--- /dev/null	(date 1727950373263)
+++ b/.idea/modules.xml	(date 1727950373263)
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ProjectModuleManager">
+    <modules>
+      <module fileurl="file://$PROJECT_DIR$/.idea/Homework Sender.iml" filepath="$PROJECT_DIR$/.idea/Homework Sender.iml" />
+    </modules>
+  </component>
+</project>
\ No newline at end of file
Index: .idea/misc.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/misc.xml b/.idea/misc.xml
new file mode 100644
--- /dev/null	(date 1727950373362)
+++ b/.idea/misc.xml	(date 1727950373362)
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="Black">
+    <option name="sdkName" value="Python 3.12 (Homework Sender)" />
+  </component>
+  <component name="ProjectRootManager" version="2" project-jdk-name="Python 3.12 (Homework Sender)" project-jdk-type="Python SDK" />
+</project>
\ No newline at end of file
Index: .idea/inspectionProfiles/profiles_settings.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/inspectionProfiles/profiles_settings.xml b/.idea/inspectionProfiles/profiles_settings.xml
new file mode 100644
--- /dev/null	(date 1727950373378)
+++ b/.idea/inspectionProfiles/profiles_settings.xml	(date 1727950373378)
@@ -0,0 +1,6 @@
+<component name="InspectionProjectProfileManager">
+  <settings>
+    <option name="USE_PROJECT_PROFILE" value="false" />
+    <version value="1.0" />
+  </settings>
+</component>
\ No newline at end of file
Index: exam_answers/EGE/2/1.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/2/1.py b/exam_answers/EGE/2/1.py
new file mode 100644
--- /dev/null	(date 1728118446940)
+++ b/exam_answers/EGE/2/1.py	(date 1728118446940)
@@ -0,0 +1,7 @@
+print('x y z w')
+for x in range(2):
+    for y in range(2):
+        for z in range(2):
+            for w in range(2):
+                if not (not(x <= y) or (x == z) or w):
+                    print(x, y, z, w)
\ No newline at end of file
Index: exam_answers/EGE/2/2.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/2/2.py b/exam_answers/EGE/2/2.py
new file mode 100644
--- /dev/null	(date 1728118710872)
+++ b/exam_answers/EGE/2/2.py	(date 1728118710872)
@@ -0,0 +1,7 @@
+print('x y z w')
+for x in range(2):
+    for y in range(2):
+        for z in range(2):
+            for w in range(2):
+                if not (not(y <= x) or (y == w) or z):
+                    print(x, y, z, w)
\ No newline at end of file
Index: exam_answers/EGE/2/3.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/2/3.py b/exam_answers/EGE/2/3.py
new file mode 100644
--- /dev/null	(date 1728118828313)
+++ b/exam_answers/EGE/2/3.py	(date 1728118828313)
@@ -0,0 +1,7 @@
+print('x y z w')
+for x in range(2):
+    for y in range(2):
+        for z in range(2):
+            for w in range(2):
+                if not (not(x <= y) or ((not w) <= (not z)) or w):
+                    print(x, y, z, w)
\ No newline at end of file
Index: exam_answers/EGE/2/4.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/2/4.py b/exam_answers/EGE/2/4.py
new file mode 100644
--- /dev/null	(date 1728118988048)
+++ b/exam_answers/EGE/2/4.py	(date 1728118988048)
@@ -0,0 +1,7 @@
+print('x y z w')
+for x in range(2):
+    for y in range(2):
+        for z in range(2):
+            for w in range(2):
+                if not (not(w <= x) or ((not z) <= (not y)) or z):
+                    print(x, y, z, w)
\ No newline at end of file
Index: exam_answers/EGE/2/5.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/2/5.py b/exam_answers/EGE/2/5.py
new file mode 100644
--- /dev/null	(date 1728119359887)
+++ b/exam_answers/EGE/2/5.py	(date 1728119359887)
@@ -0,0 +1,7 @@
+print('x y z w F')
+for x in range(2):
+    for y in range(2):
+        for z in range(2):
+            for w in range(2):
+                F = x and (y <= z) and ((not y) <= (not z == w))
+                print(x, y, z, w, int(F))
\ No newline at end of file
Index: exam_answers/EGE/2/6.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/2/6.py b/exam_answers/EGE/2/6.py
new file mode 100644
--- /dev/null	(date 1728119625407)
+++ b/exam_answers/EGE/2/6.py	(date 1728119625407)
@@ -0,0 +1,7 @@
+print('x y z w F')
+for x in range(2):
+    for y in range(2):
+        for z in range(2):
+            for w in range(2):
+                F = y and (x <= w) and ((not x) <= (not w == z))
+                print(x, y, z, w, int(F))
\ No newline at end of file
Index: exam_answers/EGE/2/7.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/2/7.py b/exam_answers/EGE/2/7.py
new file mode 100644
--- /dev/null	(date 1728119803410)
+++ b/exam_answers/EGE/2/7.py	(date 1728119803410)
@@ -0,0 +1,7 @@
+print('x y z w F')
+for x in range(2):
+    for y in range(2):
+        for z in range(2):
+            for w in range(2):
+                F = not(y <= (not(z <= w))) and ((not z) <= (not w == x))
+                print(x, y, z, w, int(F))
\ No newline at end of file
Index: exam_answers/EGE/5/1.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/5/1.py b/exam_answers/EGE/5/1.py
new file mode 100644
--- /dev/null	(date 1728118624601)
+++ b/exam_answers/EGE/5/1.py	(date 1728118624601)
@@ -0,0 +1,17 @@
+def translator(n, cc):
+    n_cc = ''
+    while n > 0:
+        n_cc += str(n % cc)
+        n //= cc
+    return n_cc[::-1]
+
+for N in range(1000, 0, -1):
+    N_4 = translator(N, 4)
+    if N % 4 == 0:
+        N_4 += N_4[-2:]
+    else:
+        N_4 += translator((N % 4) * 2, 4)
+    R = int(N_4, 4)
+    if R < 261:
+        print(N)
+        break
\ No newline at end of file
Index: exam_answers/EGE/5/2.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/5/2.py b/exam_answers/EGE/5/2.py
new file mode 100644
--- /dev/null	(date 1728118710864)
+++ b/exam_answers/EGE/5/2.py	(date 1728118710864)
@@ -0,0 +1,17 @@
+def translator(n, cc):
+    n_cc = ''
+    while n > 0:
+        n_cc += str(n % cc)
+        n //= cc
+    return n_cc[::-1]
+
+for N in range(1000, 0, -1):
+    N_4 = translator(N, 4)
+    if N % 4 == 0:
+        N_4 += N_4[-2:]
+    else:
+        N_4 += translator((N % 4) * 2, 4)
+    R = int(N_4, 4)
+    if R < 369:
+        print(N)
+        break
\ No newline at end of file
Index: exam_answers/EGE/5/3.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/5/3.py b/exam_answers/EGE/5/3.py
new file mode 100644
--- /dev/null	(date 1728118847879)
+++ b/exam_answers/EGE/5/3.py	(date 1728118847879)
@@ -0,0 +1,17 @@
+def translator(n, cc):
+    n_cc = ''
+    while n > 0:
+        n_cc += str(n % cc)
+        n //= cc
+    return n_cc[::-1]
+
+for N in range(1, 1000):
+    N_4 = translator(N, 4)
+    if N % 4 == 0:
+        N_4 += N_4[-2:]
+    else:
+        N_4 += translator((N % 4) * 2, 4)
+    R = int(N_4, 4)
+    if R >= 1025:
+        print(N)
+        break
\ No newline at end of file
Index: exam_answers/EGE/5/4.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/5/4.py b/exam_answers/EGE/5/4.py
new file mode 100644
--- /dev/null	(date 1728119022560)
+++ b/exam_answers/EGE/5/4.py	(date 1728119022560)
@@ -0,0 +1,17 @@
+def translator(n, cc):
+    n_cc = ''
+    while n > 0:
+        n_cc += str(n % cc)
+        n //= cc
+    return n_cc[::-1]
+
+for N in range(1, 1000):
+    N_4 = translator(N, 4)
+    if N % 4 == 0:
+        N_4 += N_4[-2:]
+    else:
+        N_4 += translator((N % 4) * 2, 4)
+    R = int(N_4, 4)
+    if R >= 1088:
+        print(N)
+        break
\ No newline at end of file
Index: exam_answers/EGE/5/5.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/5/5.py b/exam_answers/EGE/5/5.py
new file mode 100644
--- /dev/null	(date 1728119526115)
+++ b/exam_answers/EGE/5/5.py	(date 1728119526115)
@@ -0,0 +1,14 @@
+for N in range(1, 1000):
+    N_2 = bin(N)[2:]
+    if N % 2 == 0:
+        N_2 += '0'
+    else:
+        N_2 += '1'
+    if N_2.count('1') % 3 == 0:
+        N_2 = '11' + N_2[2:]
+    else:
+        N_2 = '10' + N_2[2:]
+    R = int(N_2, 2)
+    if R >= 26:
+        print(N)
+        break
\ No newline at end of file
Index: exam_answers/EGE/5/6.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/5/6.py b/exam_answers/EGE/5/6.py
new file mode 100644
--- /dev/null	(date 1728119625419)
+++ b/exam_answers/EGE/5/6.py	(date 1728119625419)
@@ -0,0 +1,14 @@
+for N in range(1000, 1, -1):
+    N_2 = bin(N)[2:]
+    if N % 2 == 0:
+        N_2 += '0'
+    else:
+        N_2 += '1'
+    if N_2.count('1') % 3 == 0:
+        N_2 = '11' + N_2[2:]
+    else:
+        N_2 = '10' + N_2[2:]
+    R = int(N_2, 2)
+    if R <= 37:
+        print(N)
+        break
\ No newline at end of file
Index: exam_answers/EGE/5/7.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/exam_answers/EGE/5/7.py b/exam_answers/EGE/5/7.py
new file mode 100644
--- /dev/null	(date 1728119951003)
+++ b/exam_answers/EGE/5/7.py	(date 1728119951003)
@@ -0,0 +1,8 @@
+for N in range(1, 1000):
+    N_2 = bin(N)[2:]
+    if len(N_2) % 2 == 0:
+        N_2 = N_2[0:len(N_2) // 2] + '1' + N_2[len(N_2) // 2:]
+    R = int(N_2, 2)
+    if R >= 26:
+        print(N)
+        break
\ No newline at end of file
diff --git a/test.py b/test.py
new file mode 100644
